src/d3a/models/strategy/future/strategy.py

```diff
@@ -0,0 +1,253 @@
+                final_rate=final_selling_rate,
+                fit_to_limit=True,
+                energy_rate_change_per_update=None,
+                update_interval=duration(minutes=self._update_interval),
+                rate_limit_object=max)
+
+    def event_market_cycle(self, strategy: "BidEnabledStrategy") -> None:
+        """
+        Should be called by the event_market_cycle of the asset strategy class, posts
+        settlement bids and offers on markets that do not have posted bids and offers yet
+        Args:
+            strategy: Strategy object of the asset
+
+        Returns: None
+
+        """
+        if not strategy.area.future_markets:
+            return
+        self._bid_updater.update_and_populate_price_settings(strategy.area)
+        self._offer_updater.update_and_populate_price_settings(strategy.area)
+        for time_slot in strategy.area.future_markets.market_time_slots:
+            if strategy.asset_type == AssetType.CONSUMER:
+                self._post_consumer_first_bid(strategy, time_slot)
+            elif strategy.asset_type == AssetType.PRODUCER:
+                self._post_producer_first_offer(strategy, time_slot)
+            elif strategy.asset_type == AssetType.PROSUMER:
+                self._post_prosumer_first_bid_offer(strategy, time_slot)
+            else:
+                assert False, ("Strategy %s has to be producer or consumer to be able to "
+                               "participate in the future market.", strategy.owner.name)
+
+        self._bid_updater.increment_update_counter_all_markets(strategy)
+        self._offer_updater.increment_update_counter_all_markets(strategy)
+
+    def _post_consumer_first_bid(
+            self, strategy: "BidEnabledStrategy", time_slot: DateTime) -> None:
+        required_energy_kWh = strategy.state.get_energy_requirement_Wh(time_slot) / 1000.0
+        if required_energy_kWh <= 0.0:
+            return
+        if strategy.get_posted_bids(strategy.area.future_markets, time_slot):
+            return
+        strategy.post_bid(
+            market=strategy.area.future_markets,
+            energy=required_energy_kWh,
+            price=required_energy_kWh * self._bid_updater.initial_rate[time_slot],
+            time_slot=time_slot)
+
+    def _post_producer_first_offer(
+            self, strategy: "BidEnabledStrategy", time_slot: DateTime) -> None:
+        available_energy_kWh = strategy.state.get_available_energy_kWh(time_slot)
+        if available_energy_kWh <= 0.0:
+            return
+        if strategy.get_posted_offers(strategy.area.future_markets, time_slot):
+            return
+        strategy.post_offer(
+            market=strategy.area.future_markets,
+            replace_existing=False,
+            energy=available_energy_kWh,
+            price=available_energy_kWh * self._offer_updater.initial_rate[time_slot],
+            time_slot=time_slot
+        )
+
+    def _post_prosumer_first_bid_offer(
```

Comment:
```suggestion
    def _post_prosumer_first_order(
```

Reply:
Because post_producer_first_offer and post_consumer_first_bid have different methods in their respective state classes to read the available energy for sell / buy (get_energy_requirement / get_available_energy instead of get_available_energy_to_buy/sell). Since you mentioned it though I'll try to refactor it to avoid the duplication of the post_first_bid/offer

Reply:
I guess the refactoring was not looking better, right?

Reply:
Darn, I think that github has messed up the commit diffs again and you cannot see the correct changes. The refactoring looked great, you can find it here https://github.com/gridsingularity/gsy-e/pull/1277/files#diff-bf32dee2ce4eaa87404d14ed3015f3a70bf17ee8d4d87a2132ae68cbae7e0793R170 . Essentially I completely removed the prosumer method and called post_producer_first_offer and _post_consumer_first_bid independently

Reply:
👀  weird indeed, looking into it